### DISCUSSION TITLE
Even Split

### DISCUSSION THREAD

**Discussion by eric-hemasystems** [orbitalquark/textadept#267] [Show and tell]
I use split screens a lot. Mostly vertical splits as I have a widescreen monitor. It allows me to view several files as once.

When I tried this out in Textadept, the splitting is not exactly what I want. It cuts a screen in half vs re-distributing among all the same in that row/column.

![Screen Shot 2022-09-26 at 15 31 02](https://user-images.githubusercontent.com/34583914/192363608-a0289f70-9986-4881-890d-58c331b83706.png)

Below is a work-in-progress module to make it so when I split the screen it gives each split panel an even amount of space among all views in that same row/column:

```lua
local M = {}

local function get_views(st, current_dir, size, nested)
  if st.vertical == nil then
    -- View so just return it
    return { st }
  elseif current_dir == st.vertical then
    -- View in same direction, include with list
    local a = get_views(st[1], current_dir, size, nested)
    local b = get_views(st[2], current_dir, size, nested)
    return table.move(b, 1, #b, #a + 1, a)
  else
    -- View in different direction. Note for nested processing
    table.insert(nested, st)
    return {}
  end
end

local function resize(st, current_dir, size)
  -- Gather up all views in same col/row and note nested ones that switch direction
  local nested = {}
  local views = get_views(st, current_dir, size, nested)

  -- Calculate size for this col/row
  local max = st.vertical and size[1] or size[2]
  local per = math.floor(max / (#views + #nested))

  -- Assign new caculated size to each view in this col/row
  for _, view in ipairs(views) do
    view.size = per
  end

  -- Figure out how much the nested cols/rows have to work with
  local remaining
  if current_dir then
    remaining = { size[1] - per, size[2] }
  else
    remaining = { size[1], size[2] - per }
  end

  -- Process nested col/rows
  for _, nest in ipairs(nested) do
    resize(nest, nest.vertical, remaining)
  end
end

function M.even_split()
  local st = ui.get_split_table()
  resize(st, st.vertical, ui.size)
end

events.connect(events.VIEW_NEW, function()
  -- QUESTION: Why is the timeout needed?
  timeout(0.01, M.even_split)
end)

return M
```

With this now when I hit `Cmd-V` a few times I get this which is more useful:

![Screen Shot 2022-09-26 at 15 34 43](https://user-images.githubusercontent.com/34583914/192364160-6d9e1ae4-e4ab-4ffc-b53f-c43f85e9cf1e.png)

It should also work on horizontal splits as well as nested stuff like this:

![Screen Shot 2022-09-26 at 15 37 55](https://user-images.githubusercontent.com/34583914/192364718-40142539-4ecd-4e60-ae93-1de5c2a6a130.png)

I call this a work-in-progress for a few reasons:

1. Right now it will re-align all views. I'm wondering if I instead want it to just re-align the views in the current row/column in case other rows/columns have been hand-adjusted.
2. It will briefly show at the undesired size before correcting. I'm guessing this relates to my timeout workaround although I'm not sure what the solution is.
3. There is some funkiness that happens when the split view is not the last on one the row/column. For example if you just `Cmd-V` two times in a row you get your three views all of equal size. But if you instead do `Cmd-V` then change the focus to the first view and split that one it ends up setting them to the wrong sizes.

I'm sure I'll refine this a bit over time but wanted to at least share my early draft of this behavior.

**Comment by eric-hemasystems** [comment]
Also i would love to trigger the function to run when the Textadept window is resized but I don't see an event for that. I see I can get the windows width/height so perhaps I could poll and adjust when it changes....

**Comment by orbitalquark** [comment]
For (2), it's probably because the view hasn't yet been painted, so size information is not yet available. Try calling `ui.update()` before asking for and acting on sizes. You shouldn't need the timeout anymore.

**Comment by orbitalquark** [comment]
You are correct that there is no event on window resize. Polling is probably the way to go for now if you want it to happen automatically.

**Reply by eric-hemasystems** [reply]
I also realized that I'm not redistributing when I close a view (i.e. when I go from 3 views to 2 it doesn't resize from 33% to 50%). I see an `unsplit` method but it doesn't indicate a event like the `split` does. Sounds like I will need polling to handle for that as well but just want to confirm I'm not missing some other way to know when the split table has changed.

**Comment by orbitalquark** [comment]
There is no event when a view is unsplit.

**Comment by eric-hemasystems** [comment]
I'm starting to wonder if this can really be solved with the API. I misunderstood what `size` on the view is. It works if I always split on the last view in a row/column. But if I split on earlier views I get problems.

For example if I split on the last view always then after two splits I have:

```
View 1 - 1146px
Split
  View 2 - 1146px
  View 3 - 1146px
```

But if I split in half and then split the first view to make them even the size should be:

```
Split
  View 1 - 1146px
  View 2 - 1146px
View 3 - 2292px
```

I was thinking I could adjust my process to account for this but then came across this layout:

```
Split 1
  View 1 - 860
  View 2 - 860
Split 2
  View 3 - 860
  View 4 - 860
```

So I split in half and then on each half split again. I can of course set the size of each view to 860. But there is no variable that I can see to see to set the size between the two splits. I.E. you can have all views with 860 and still have multiple possible layouts depending on the position of the top-level split.

**Comment by orbitalquark** [comment]
You've identified a limitation of the API. `view.size` only applies to the view's parent split (if any). With your dual-split arrangement, each `view.size` only applies to the "sub-splits", with no way to reference or set the parent/top-level split.

**Reply by eric-hemasystems** [reply]
I think I may be able to work around this actually by combining this code with another customization I was planning related to views.

I actually don't really split other than on the last view anyway. I was just trying to make this functionality general purpose (support multiple splitting workflows). Sounds like I'll just bundle it with my other customization rather than making it general purpose which avoids the issue.

Thanks for the pointers though.

**Comment by orbitalquark** [comment]
You are correct that there is no event on window resize. Polling is probably the way to go for now if you want it to happen automatically.
